lib/RunTests.js

```diff
@@ -0,0 +1,274 @@
+// @flow
+
+const chalk = require('chalk');
+const chokidar = require('chokidar');
+const path = require('path');
+const packageInfo = require('../package.json');
+const Compile = require('./Compile');
+const ElmJson = require('./ElmJson');
+const FindTests = require('./FindTests');
+const Generate = require('./Generate');
+const Project = require('./Project');
+const Report = require('./Report');
+const Supervisor = require('./Supervisor');
+
+void Report;
+
+// Incorporate the process PID into the socket name, so elm-test processes can
+// be run parallel without accidentally sharing each others' sockets.
+//
+// See https://github.com/rtfeldman/node-test-runner/pull/231
+// Also incorporate a salt number into it on Windows, to avoid EADDRINUSE -
+// see https://github.com/rtfeldman/node-test-runner/issues/275 - because the
+// alternative approach of deleting the file before creating a new one doesn't
+// work on Windows. We have to let Windows clean up the named pipe. This is
+// essentially a band-aid fix. The alternative is to rewrite a ton of stuff.
+function getPipeFilename(runsExecuted /*: number */) /*: string */ {
+  return process.platform === 'win32'
+    ? `\\\\.\\pipe\\elm_test-${process.pid}-${runsExecuted}`
+    : `/tmp/elm_test-${process.pid}.sock`;
+}
+
+function infoLog(
+  report /*: typeof Report.Report */,
+  msg /*: string */
+) /*: void */ {
+  if (report === 'console') {
+    console.log(msg);
+  }
+}
+
+function clearConsole(report /*: typeof Report.Report */) {
+  if (report === 'console') {
+    process.stdout.write(
+      process.platform === 'win32' ? '\x1B[2J\x1B[0f' : '\x1B[2J\x1B[3J\x1B[H'
+    );
+  }
+}
+
+function diffArrays/*:: <T> */(
+  from /*: Array<T> */,
+  to /*: Array<T> */
+) /*: { added: Array<T>, removed: Array<T> } */ {
+  return {
+    added: to.filter((item) => !from.includes(item)),
+    removed: from.filter((item) => !to.includes(item)),
+  };
+}
+
+function delay(ms /*: number */) /*: Promise<void> */ {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
+}
+
+const Queue /*: Array<{
+  event: 'added' | 'changed' | 'removed',
+  filePath: string,
+}> */ = [];
+void Queue;
+
+function watcherEventMessage(queue /*: typeof Queue */) /*: string */ {
+  const suffix = '. Rebuilding!';
+
+  const filePaths = Array.from(new Set(queue.map(({ filePath }) => filePath)));
+  if (filePaths.length === 1) {
+    const { event, filePath } = queue[0];
+    return `${filePath} ${event}${suffix}`;
+  }
+
+  const events = Array.from(new Set(queue.map(({ event }) => event))).sort();
+  return `${filePaths.length} files ${events.join('/')}${suffix}`;
+}
+
+function runTests(
```

Comment:
This function is the `test` function plus the `runTests` function from elm-test.js. I started by merging those two into one – they were a very arbitrary split, where the last thing `test` did was calling `runTests`. I find it much easier to understand when they are one.

Then there is some new code for the improved queuing of tests and how the watcher is set up.